### ISSUE TITLE
High response latency for bi-directional streaming when multiple responses are sent from host 

### ISSUE THREAD

**Issue by sreejit-a** [grpc/grpc#23232]
### What version of gRPC and what language are you using?
C# .Net 4.6.2 - Grpc Nuget Version 2.29.0

### What operating system (Linux, Windows,...) and version?
Windows

### What did you do?

We have a bi-directional grpc streaming application. The client sends a request to host. Host can either process the request and send one and final response back to the client and shutdown the stream or sends for additional request and waits for reply from client. 

We notice that in cases where server completes with only response, the latency is very minimal. But, when host sends request and does not close the stream immediately, we notice high latency time for that response.

The code below is a repro for this issue - SingleExecute call writes a single response and completes immediately. DoubleExecute call writes two responses. DoubleExecute also has a delay of 5 ms to simulate some server work. 

SingleExecute times are almost always 0ms when both server and client are on the same machine. DoubleExecute times always have a delay of 12-18 ms instead of having a latency around 5 ms.

If we remove delay in DoubleExecute, then response times for DoubleExecute is back to 0ms. Actual application has meaningful work to be done before it can send a response back requesting for more information so this delay cannot be avoided.

Proto file
```
syntax = "proto3";

service TestService {
	rpc SingleResponseExecute(stream Request) returns (stream Response) {}
	rpc DoubleResponseExecute(stream Request) returns (stream Response) {}
}

message Request {
}

message Response {
}
```

Server code 

```
 public class TestServer : TestService.TestServiceBase
    {
        public override async Task SingleResponseExecute(IAsyncStreamReader<Request> requestStream, IServerStreamWriter<Response> responseStream, ServerCallContext context)
        {
            await responseStream.WriteAsync(new Response()).ConfigureAwait(false);
        }

        public override async Task DoubleResponseExecute(IAsyncStreamReader<Request> requestStream, IServerStreamWriter<Response> responseStream, ServerCallContext context)
        {
            try
            {
                await Task.Delay(5).ConfigureAwait(false); // Simulate some server work
                await responseStream.WriteAsync(new Response()).ConfigureAwait(false);
                await responseStream.WriteAsync(new Response()).ConfigureAwait(false);
            }
            catch(Exception ex)
            {
                Console.WriteLine(ex.ToString());
                throw;
            }
        }
    }
```

Client Code 
```
for (int i = 0; i < 100; i++)
            {
                stopwatch.Restart();
                long firstResponseTime = 0;
                using (var doubleExecute = client.DoubleResponseExecute(deadline: DateTime.UtcNow.AddSeconds(3600)))
                {
                    await doubleExecute.RequestStream.WriteAsync(new Request()).ConfigureAwait(false);
                    while (await doubleExecute.ResponseStream.MoveNext().ConfigureAwait(false))
                    {
                        firstResponseTime = stopwatch.ElapsedMilliseconds;
                        var response = doubleExecute.ResponseStream.Current;
                        //await doubleExecute.RequestStream.WriteAsync(new Request()).ConfigureAwait(false);
                    }
                    await doubleExecute.RequestStream.CompleteAsync();
                }
                stopwatch.Stop();
                Console.WriteLine($"Double execute Attempt {i}: First Reponse: {firstResponseTime} Total Time: {stopwatch.ElapsedMilliseconds}");
            }
```

### What did you expect to see?
When running SingleExecute that sends only one response and shuts the stream, the time taken is always close to 0, When running DoubleExecute method that writes two response streams - average time should be around the 5ms delay (which was added to simulate server work), but instead, latency times are around 12-18 ms.

### What did you see instead?
When server does some work and then attempts send response, we are noticing high latency - around 12-18 ms even though both server and client are on the same machine. The latency is higher when server and client are on different machines on the same azure regions and thus impacting even more.

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.

### Anything else we should know about your project / environment?
WriteAsync do not have any buffer hint set.

**Comment by sreejit-a** [comment]
I enabled Grpc "Info" log on both the server and client, and we see the following on the server side - 
I can see there is a 5ms delay in the second log line which is expected (as there is a Task.Delay in the server code)

However, the subsequent **grpc_call_start_batch** log starts only after about 10-11 ms - which is correlating and contributing to the delay we are seeing.


I0617 09:42:27.785838 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\surface\call_details.cc:37: grpc_call_details_destroy(cd=00B3BF78)

I0617 **09:42:27.790823** 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\iomgr\tcp_windows.cc:187: TCP:00B13DF8 on_read
I0617 09:42:27.790823 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\iomgr\tcp_windows.cc:216: READ 00B13DF8 (peer=ipv4:127.0.0.1:5518): 00 00 05 00 00 00 00 00 01 00 00 00 00 00 '..............'
I0617 09:42:27.790823 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\iomgr\timer_generic.cc:468: TIMER 00B179D0: CANCEL pending=true
I0617 09:42:27.791821 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\iomgr\tcp_windows.cc:253: TCP:00B13DF8 win_read
I0617 09:42:27.791821 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\iomgr\timer_generic.cc:365: TIMER 00B179D0: SET 7270475 now 70475 call 00B17990[799E6EE0]
I0617 09:42:27.791821 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\iomgr\timer_generic.cc:402:   .. add to shard 4 with queue_deadline_cap=71078 => is_first_timer=false


I0617 **09:42:27.802792** 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\surface\call.cc:1967: **grpc_call_start_batch**(call=0504E190, ops=00EAEBCC, nops=2, tag=00B47070, reserved=00000000)
I0617 09:42:27.802792 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\surface\call.cc:1565: ops[0]: SEND_MESSAGE ptr=05E3E8C0
I0617 09:42:27.802792 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\surface\call.cc:1565: ops[1]: SEND_INITIAL_METADATA(nil)
I0617 09:42:27.803789 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\surface\call.cc:626: OP[server:0504E6B0]:  SEND_INITIAL_METADATA{} SEND_MESSAGE:flags=0x00000000:len=0
I0617 09:42:27.803789 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\channel\channel_stack.cc:226: OP[message_size:0504E6BC]:  SEND_INITIAL_METADATA{} SEND_MESSAGE:flags=0x00000000:len=0
I0617 09:42:27.803789 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\channel\channel_stack.cc:226: OP[deadline:0504E6C8]:  SEND_INITIAL_METADATA{} SEND_MESSAGE:flags=0x00000000:len=0
I0617 09:42:27.803789 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\channel\channel_stack.cc:226: OP[http-server:0504E6D4]:  SEND_INITIAL_METADATA{} SEND_MESSAGE:flags=0x00000000:len=0
I0617 09:42:27.803789 0 T:\src\github\grpc\workspace_csharp_ext_windows_x86\src\core\lib\channel\channel_stack.cc:226: OP[message_decompress:0504E6E0]:  SEND_INITIAL_METADATA{key=3a 73 74 61 74 75 73 ':status' value=32 30 30 '200', key=63 6f 6e 74 65 6e 74 2d 74 79 70 65 'content-type' value=61 70 70 6c 69 63 61 74 69 6f 6e 2f 67 72 70 63 'application/grpc'} SEND_MESSAGE:flags=0x00000000:len=0

This delay should be as minimal as possible.

**Comment by sreejit-a** [comment]
@jtattermusch - Is there any configuration that we need set up to prevent this delay?

**Comment by sreejit-a** [comment]
Hi @jtattermusch - We perform multiple streaming responses and we are seeing these latencies add up significantly impacting overall performance. 

Any guidance here will help us. Let me know if you need any further information

**Comment by jtattermusch** [comment]
@sreejit-a  can you provide a reproduction in form of a ready-to-run solution (e.g. in a github repository) - that would make it easier to reproduce.

Also, what windows version are you using?

**Comment by sustephen** [comment]
Ran into the similar issue with c++ async bi-directional streaming as well. In our case the delay can be as high as 2 mins. The problem only happens on system that is relatively busy with other socket IOs in the same process. ss is reporting the packet in the RecvQ, however it seems grpc just not reading the packet from socket.  
I workaround the issue by calling AsyncNext() with deadline of 1 msec instead of 0.

**Comment by saiprabhur** [comment]
any fix found for the issue [sreejit-a](https://github.com/sreejit-a)

**Comment by tonydnewell** [comment]
@sreejit-a I've tried to reproduce what you are seeing using the example code you have provided.
Some observations:

- the server example code does not consume the client requests (does not read the requestStream). I've fixed this in my test.
- you only have a task.Delay(5) in DoubleResponseExecute. If I add the same delay to SingleResponseExecute or remove it from DoubleResponseExecute then results I get are identical

The documentation for Task.Delay states:

> This method depends on the system clock. This means that the time delay will approximately equal the resolution of the system clock if the millisecondsDelay argument is less than the resolution of the system clock, which is approximately 15 milliseconds on Windows systems.

See https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.delay?view=netstandard-1.4

So the larger than expected latency times are due to resolution of the system clock on Windows.
